Added full typings for vue2-geocoder

simon-lanf opened the pull request:
If you have any questions feel free to ask.

simon-lanf committed:
Added full typings for vue2-geocoder

simon-lanf committed:
optional location object fields

simon-lanf committed:
optional toAddressString param

simon-lanf committed:
Added response type

simon-lanf committed:
Brought back source comments for enums union types

simon-lanf committed:
Allow async functions as callback

simon-lanf committed:
Fixed promise handling

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+interface Vue2GeocoderLatLngObj {
+    lat: number;
+    lng: number;
+}
```
This matches the setup for the `latLngObj` as outlined in the [README](https://github.com/PDERAS/vue2-geocoder/blob/master/README.md)

```js
var latLngObj = {
    lat: 37.421512,
    lng: -122.084101
}
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    state?: string;
+    zip_code?: string;
+    country?: string;
+}
```
This matches the setup for the `addressObj` as outlined in the [README](https://github.com/PDERAS/vue2-geocoder/blob/master/README.md)

```js
var addressObj = {
    address_line_1: '1600 Amphitheatre Parkway',
    address_line_2: '',
    city:           'Mountain View',
    state:          'CA',               // province also valid
    zip_code:       '94043',            // postal_code also valid
    country:        'United States'
}
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    lng: number;
+}
+
+type Vue2GeocoderModes = 'lat-lng' | 'address';
```
This matches the setup for the `defaultMode` option as outlined in the [README](https://github.com/PDERAS/vue2-geocoder/blob/master/README.md)

```js
defaultMode:        'address', // or 'lat-lng'
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'YE'
+    | 'ZM'
+    | 'ZW'
+    | 'AX';
```
I double checked all the country codes to the [developer documentation](https://developers.google.com/maps/coverage) and it all matches 👍.

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'vi'
+    | 'id'
+    | 'zu'
+    | 'it';
```
I double checked the rest of all the language codes to the [developer documentation](https://developers.google.com/maps/faq#languagesupport) and it all matches 👍.

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+
+// source: https://developers.google.com/maps/faq#languagesupport
+type Vue2GeocoderLanguageCode =
+    | 'ja'
```
Looks like we are missing just one

```suggestion
type Vue2GeocoderLanguageCode =
    | 'af'
    | 'ja'
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'it';
+
+// source: https://developers.google.com/maps/documentation/geocoding/overview#Types
+type AddressComponentType =
+    | 'street_address'
+    | 'route'
+    | 'intersection'
+    | 'political'
+    | 'country'
+    | 'administrative_area_level_1'
+    | 'administrative_area_level_2'
+    | 'administrative_area_level_3'
+    | 'administrative_area_level_4'
+    | 'administrative_area_level_5'
+    | 'colloquial_area'
+    | 'locality'
+    | 'sublocality'
+    | 'neighborhood'
+    | 'premise'
+    | 'subpremise'
+    | 'plus_code'
+    | 'postal_code'
+    | 'natural_feature'
+    | 'airport'
+    | 'park'
+    | 'point_of_interest'
+    | 'floor'
+    | 'establishment'
+    | 'landmark'
+    | 'point_of_interest'
+    | 'parking'
+    | 'post_box'
+    | 'postal_town'
+    | 'room'
+    | 'street_number'
+    | 'bus_station';
```
I double checked all the rest of the address component types to the [developer documentation](https://developers.google.com/maps/documentation/geocoding/overview#Types) and it all matches 👍.

codymoorhouse replied:
Just missing two types as per the docs

> `bus_station`, `train_station` and `transit_station` indicate the location of a bus, train or public transit stop.

```suggestion
    | 'street_number'
    | 'bus_station'
    | 'train_station'
    | 'transit_station';
```

codymoorhouse replied:
I double checked all the rest of the address component types to the [developer documentation](https://developers.google.com/maps/documentation/geocoding/overview#Types) and it all matches 👍.

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'street_number'
+    | 'bus_station';
+
+// source: https://developers.google.com/maps/documentation/geocoding/overview#results
+type LocationType =
+    | 'ROOFTOP'
+    | 'RANGE_INTERPOLATED'
+    | 'GEOMETRIC_CENTER'
+    | 'APPROXIMATE';
```
I double checked all the rest of the location types to the [developer documentation](https://developers.google.com/maps/documentation/geocoding/overview#results) and it all matches 👌

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'GEOMETRIC_CENTER'
+    | 'APPROXIMATE';
+
+interface AddressComponent {
+    long_name: string;
+    short_name: string;
+    types: AddressComponentType[];
+}
```
This looks good according to the docs

<img width="422" alt="Screen Shot 2020-11-18 at 6 08 21 PM" src="https://user-images.githubusercontent.com/10747839/99607468-1c022500-29c9-11eb-905d-df9701379d49.png">

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    types: AddressComponentType[];
+}
+
+interface Bound {
+    lat: number;
+    lng: number;
+}
+
+interface Bounds {
+    northeast: Bound;
+    southwest: Bound;
+}
+
+interface Location {
+    lat: number;
+    lng: number;
+}
+
+interface Viewport {
+    northeast: Bound;
+    southwest: Bound;
+}
+
+interface Geometry {
+    bounds: Bounds;
+    location: Location;
+    location_type: LocationType;
+    viewport: Viewport;
+}
```
Viewport and Bounds are the same mappings, we could reuse it or we could keep it separate. What I mean is we could remove Viewport and set the key as the `Bounds` type as well, but keeping it separate makes sense to me

e.i.
```ts
interface Geometry {
    bounds: Bounds;
    location: Location;
    location_type: LocationType;
    viewport: Bounds;
}
```

How the API result looks for reference:

<img width="397" alt="Screen Shot 2020-11-18 at 6 10 53 PM" src="https://user-images.githubusercontent.com/10747839/99607646-6daaaf80-29c9-11eb-98f2-1e734be9fd48.png">

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    | 'it';
+
+// source: https://developers.google.com/maps/documentation/geocoding/overview#Types
+type AddressComponentType =
```
> The following types are supported and returned by the geocoder in both the address type and address component type arrays:

Lets rename this to just `AddressType`, then we can use it in the results for the `types[]` key as well. We will have to also update it in the `AddressComponent` type

```suggestion
type AddressType =
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+    formatted_address: string;
+    geometry: Geometry;
+    place_id: string;
+    types: string[];
```
Once we update the `AddressComponentType` type to `AddressType` it makes sense to do this

```suggestion
    types: AddressType[];
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+
+export interface Result {
+    address_components: AddressComponent[];
+    formatted_address: string;
```
> `postcode_localities[]` is an array denoting all the localities contained in a postal code. This is only present when the result is a postal code that contains multiple localities.

We should probably support the other possible result keys as well (https://developers.google.com/maps/documentation/javascript/geocoding)

```ts
postcode_localities[]: string,
```

> `partial_match` indicates that the geocoder did not return an exact match for the original request

```ts
partial_match: boolean
```

simon-lanf replied:
I'll try to do it tomorrow, otherwise you can make an edit here if I forget.

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+interface AddressComponent {
+    long_name: string;
+    short_name: string;
+    types: AddressComponentType[];
```
We should add this into this interface ([see here](https://developers.google.com/maps/documentation/javascript/geocoding))

```ts
postcode_localities[]: string,
```

codymoorhouse commented on index.d.ts:
```diff
@@ -0,0 +1,490 @@
+}
+
+// source: https://developers.google.com/maps/documentation/geocoding/overview#StatusCodes
+type Statuses =
+    | 'ZERO_RESULTS'
+    | 'OK'
+    | 'OVER_DAILY_LIMIT'
+    | 'OVER_QUERY_LIMIT'
+    | 'REQUEST_DENIED'
+    | 'INVALID_REQUEST'
+    | 'UNKNOWN_ERROR';
```
I double checked all statuses to the [developer documentation](https://developers.google.com/maps/documentation/geocoding/overview#StatusCodes) and it all matches 👌

codymoorhouse reviewed (commented):
This is cool, not completely sure how to test all this, and I will leave that up to PDERAS (_as I do not maintain this package anymore_). We should probably include some examples / usages in the README somehow. Just thought I would help and offer what I noticed at a code / accuracy level. Great work Simon!

simon-lanf committed:
Add missing language code

Co-authored-by: Cody Moorhouse <cody.moorhouse@icloud.com>

simon-lanf committed:
Renamed addresstype to reuse it

Co-authored-by: Cody Moorhouse <cody.moorhouse@icloud.com>

simon-lanf committed:
Use AddressType everywhere

Co-authored-by: Cody Moorhouse <cody.moorhouse@icloud.com>

simon-lanf commented:
Thank you very much @codymoorhouse for this code review, this is really appreciated. I'll add something to the readme to explain how I use it.

simon-lanf committed:
Wrapped in a module and augmenting Vue

More info: 
https://vuejs.org/v2/guide/typescript.html#Augmenting-Types-for-Use-with-Plugins

simon-lanf commented:
Now the module is wrapped and Vue is automatically augmented with the types for the plugin with no effort at all. I used these example from the docs: https://vuejs.org/v2/guide/typescript.html#Augmenting-Types-for-Use-with-Plugins

This is the first time I do types for a Vue plugin let me know if I missed something. I still have to add some of @codymoorhouse ideas into the code.

simon-lanf commented:
If @codymoorhouse or @PDERAS want to test it, just put the index.d.ts file in node_modules/@pderas/vue2-geocoder inside a vue project using typescript, import the module and check the completion in VSCode.
